extensions/coap/nanofi/coap_functions.c

```diff
@@ -0,0 +1,175 @@
+
+      *ctx = coap_new_context(0x00);
+
+      *session = coap_new_client_session(*ctx, &addr, dst_addr, proto);
+      if (*ctx && *session) {
+        freeaddrinfo(result);
+        return 0;
+      }
+    }
+  }
+
+  fprintf(stderr, "no context available for interface '%s'\n", node);
+
+  freeaddrinfo(result);
+  return -1;
+}
+
+struct coap_pdu_t *create_request(struct coap_context_t *ctx,struct coap_session_t *session,coap_optlist_t **optlist, unsigned char code, coap_str_const_t *ptr) {
+  coap_pdu_t *pdu;
+
+  if (!(pdu = coap_new_pdu(session)))
+    return NULL;
+
+  pdu->type = COAP_MESSAGE_CON;
+  pdu->tid = coap_new_message_id(session);
+  pdu->code = code;
+
+  if (optlist){
+    coap_add_optlist_pdu(pdu, optlist);
+  }
+
+  int flags = 0;
+  coap_add_data(pdu, ptr->length, ptr->s);
+  return pdu;
+}
+
+int coap_event(struct coap_context_t *ctx, coap_event_t event, struct coap_session_t *session){
+    if (event == COAP_EVENT_SESSION_FAILED && global_ptrs.received_error){
+            global_ptrs.received_error(receiver, ctx, -1);
+    }
+    return 0;
+}
+
+void no_acknowledgement(struct coap_context_t *ctx, coap_session_t *session, coap_pdu_t *sent, coap_nack_reason_t reason, const coap_tid_t id){
+  if (global_ptrs.received_error){
+          global_ptrs.received_error(receiver, ctx, -1);
+  }
+}
+
+void response_handler(struct coap_context_t *ctx, struct coap_session_t *session, coap_pdu_t *sent, coap_pdu_t *received, const coap_tid_t id) {
+  unsigned char* data;
+  size_t data_len;
+  coap_opt_iterator_t opt_iter;
+  coap_opt_t * block_opt = coap_check_option(received, COAP_OPTION_BLOCK1, &opt_iter);
+  if (block_opt) {
+  } else {
+    if (!global_ptrs.data_received){
+      return;
+    }
+
+    if (COAP_RESPONSE_CLASS(received->code) == 2 || received->code == COAP_RESPONSE_400) {
+      if (coap_get_data(received, &data_len, &data)) {
+        if (global_ptrs.data_received)
+        global_ptrs.data_received(receiver, ctx, received->code, data, &data_len);
+      }
+    }
+    else{
+      if (global_ptrs.received_error)
+        global_ptrs.received_error(receiver, ctx, received->code);
+    }
+  }
+
+}
+
+int resolve_address(const struct coap_str_const_t *server, struct sockaddr *dst) {
+  struct addrinfo *res, *ainfo;
+  struct addrinfo hints;
+  static char addrstr[256];
+  int error, len = -1;
+
+  memset(addrstr, 0, sizeof(addrstr));
+  if (server->length)
+    memcpy(addrstr, server->s, server->length);
+  else
+    memcpy(addrstr, "127.0.0.1", 9);
+
+  memset((char *) &hints, 0, sizeof(hints));
+  hints.ai_socktype = SOCK_DGRAM;
+  hints.ai_family = AF_UNSPEC;
+
+  error = getaddrinfo(addrstr, NULL, &hints, &res);
+
+  if (error != 0) {
+    return error;
+  }
+
+  for (ainfo = res; ainfo != NULL; ainfo = ainfo->ai_next) {
+    switch (ainfo->ai_family) {
+      case AF_INET6:
+      case AF_INET:
+        len = ainfo->ai_addrlen;
+        memcpy(dst, ainfo->ai_addr, len);
+        freeaddrinfo(res);
```

Comment:
A break would be enough here, free and return happens at the end of the function.